schema: fix describe of indexes on collections

Jadw1 opened the pull request:
If the index was created on collection (both frozen or not), its description wasn't a correct create statement.
This patch fixes the bug and includes functions like `full()`, `keys()`, `values()`, ... used to create index on collections.

Fixes https://github.com/scylladb/scylladb/issues/19278

Jadw1 commented:
@nyh Can you plase take a look. I don't know index code, so I'm not sure my solution is correct.

The new test will fail because I some questions regarding indexes and theirs description:
Let's say you have a table:
```
CREATE TABLE ks1.tt (
    a int,
    b list<int>,
    c set<int>,
    PRIMARY KEY (a)
)
```

- Is index `CREATE INDEX idx_b ON ks1.tt(b)` the exactly the same as `CREATE INDEX idx_b ON ks1.tt(values(b))`?
Because creating index the first way adds `values()` function to index option in `system_schema.indexes`:
```
 keyspace_name | table_name | index_name | kind       | options
---------------+------------+------------+------------+-------------------------
           ks1 |         tt |      idx_b | COMPOSITES | {'target': 'values(b)'}
```

- When I'm creating index `CREATE INDEX idx_c ON ks1.tt(c)`, it's target in indexes table is `keys(c)`
```
 keyspace_name | table_name | index_name | kind       | options
---------------+------------+------------+------------+-------------------------
           ks1 |         tt |      idx_c | COMPOSITES |   {'target': 'keys(c)'}
```
But it's not possible to create a index on keys on non-map column
```
cqlsh:ks1> create index idx_c on tt(keys(c));
InvalidRequest: Error from server: code=2200 [Invalid query] message="Cannot create secondary index on keys of column c with non-map type"
```
Is this a bug or intentional design and I should print `tt(c)` instead of `tt(keys(c))` in describe?

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9828/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9828/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: cql-pytest/test_describe
#### Failed Tests (60/1050):
- [test_desc_index_on_collections.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9828/testReport/junit/%28root%29/test_describe/test_desc_index_on_collections_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_desc_index_on_collections.debug.1)
- [test_desc_index_on_collections.debug.10](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9828/testReport/junit/%28root%29/test_describe/test_desc_index_on_collections_debug_10) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_desc_index_on_collections.debug.10)
- [test_desc_index_on_collections.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9828/testReport/junit/%28root%29/test_describe/test_desc_index_on_collections_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_desc_index_on_collections.debug.2)
- [test_desc_index_on_collections.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9828/testReport/junit/%28root%29/test_describe/test_desc_index_on_collections_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_desc_index_on_collections.debug.3)
- [test_desc_index_on_collections.debug.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9828/testReport/junit/%28root%29/test_describe/test_desc_index_on_collections_debug_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_desc_index_on_collections.debug.4)
- [test_desc_index_on_collections.debug.5](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9828/testReport/junit/%28root%29/test_describe/test_desc_index_on_collections_debug_5) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_desc_index_on_collections.debug.5)
- [test_desc_index_on_collections.debug.6](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9828/testReport/junit/%28root%29/test_describe/test_desc_index_on_collections_debug_6) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_desc_index_on_collections.debug.6)
- [test_desc_index_on_collections.debug.7](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9828/testReport/junit/%28root%29/test_describe/test_desc_index_on_collections_debug_7) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_desc_index_on_collections.debug.7)
- [test_desc_index_on_collections.debug.8](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9828/testReport/junit/%28root%29/test_describe/test_desc_index_on_collections_debug_8) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_desc_index_on_collections.debug.8)
- [test_desc_index_on_collections.debug.9](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9828/testReport/junit/%28root%29/test_describe/test_desc_index_on_collections_debug_9) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_desc_index_on_collections.debug.9)
#### Build Details:
- Duration: 1 hr 1 min
- Builder: spider6.cloudius-systems.com

nyh commented:
>     * Is index `CREATE INDEX idx_b ON ks1.tt(b)` the exactly the same as `CREATE INDEX idx_b ON ks1.tt(values(b))`?
>       Because creating index the first way adds `values()` function to index option in `system_schema.indexes`:

Yes.
Check out docs/design-notes/secondary_index for some documentation of this area. See also the test suite (test/cql-pytest/test_secondary_index.py, which I wrote myself)  which verifies that `values(m)` and `m` is the same/
The pull request introducing collection indexing is https://github.com/scylladb/scylladb/pull/10841 - I didn't write most of the code but I fixed a lot of bugs before merging it.

> ```
>  keyspace_name | table_name | index_name | kind       | options
> ---------------+------------+------------+------------+-------------------------
>            ks1 |         tt |      idx_b | COMPOSITES | {'target': 'values(b)'}
> ```
> 
>     * When I'm creating index `CREATE INDEX idx_c ON ks1.tt(c)`, it's target in indexes table is `keys(c)`

Wow, what a mess :-(
I see it in cql3/statements/create_index_statement.cc:

```
        case index_target::target_type::collection_values:
            if (cd.type->is_map() || cd.type->is_list()) {
                target.type = index_target::target_type::collection_values;
            } else if (cd.type->is_set()) {
                target.type = index_target::target_type::keys;
            } else {
```

Introduced in commit 997682ed72e
This commits suggests that the values of the set are, internally, its keys - which is technically correct, which is why the target is "rewritten" to be "keys".

This means that when you want to print out this "target" back to CQL, you need to undo this rewrite_target_for_collection() code - if you see a set with target "keys" you need to remove that "keys" word and leave just the column name.

> Is this a bug or intentional design and I should print `tt(c)` instead of `tt(keys(c))` in describe?

So, it **appears** to be intentional, and yes, I think you need to remove the "keys()" in the describe.
You should write a thorough test that tries all the collection types and all the legal keys/entries/values combinations (see my test test_index_collection_wrong_type for all the legal combinations) to make sure you got it all right :-(

nyh commented on test/cql-pytest/test_describe.py:
```diff
@@ -447,6 +447,24 @@ def test_desc_index(cql, test_keyspace):
+            ("idx_c_entries",   f"CREATE INDEX idx_c_entries ON {tbl}(entries(c))"),
+            ("idx_c_keys",      f"CREATE INDEX idx_c_keys ON {tbl}(keys(c))"),
+            ("idx_c_values",    f"CREATE INDEX idx_c_values ON {tbl}(values(c))"),
+            ("idx_d",           f"CREATE INDEX idx_d ON {tbl}(d)"),    # TODO: this index will fail the assert becuase its target is "values(d)"
```
Yes, for list, indexing `d` and `values(d)` is exactly the same thing (the target will always be written as `values(d)` and that's what you can expect.

piodul replied:
Typo: `becuase` -> `because`

nyh commented on test/cql-pytest/test_describe.py:
```diff
@@ -447,6 +447,24 @@ def test_desc_index(cql, test_keyspace):
+            ("idx_c_keys",      f"CREATE INDEX idx_c_keys ON {tbl}(keys(c))"),
+            ("idx_c_values",    f"CREATE INDEX idx_c_values ON {tbl}(values(c))"),
+            ("idx_d",           f"CREATE INDEX idx_d ON {tbl}(d)"),    # TODO: this index will fail the assert becuase its target is "values(d)"
+            ("idx_e",           f"CREATE INDEX idx_e ON {tbl}(e)"),    # TODO: this index will fail the assert becuase its target is "keys(e)"
```
As I explained in the discussion, this is a bizarre implementation detail - the "target" in this case will be `key(e)` but it is keys only in some internal sense and not the CQL sense. You need to translatet his target of `keys(e)` back to `values(e)` (or just `e`) when printing it out to CQL.

nyh commented on schema/schema.cc:
```diff
@@ -838,7 +839,25 @@ std::ostream& schema::describe(replica::database& db, std::ostream& os, bool wit
+                        auto base_schema = db.find_schema(view_info()->base_id());
+                        auto index_metadata = base_schema->all_indices().at(secondary_index::index_name_from_table_name(cf_name()));
+                        auto target = index_metadata.options().at(cql3::statements::index_target::target_option_name);
+                        auto index_target_name = secondary_index::target_parser::get_target_column_name_from_string(target);
```
As discussed above, you need to undo here the mess that `create_index_statement::rewrite_target_for_collection() did. The only bad thing it did (I think) is to add the "keys()" when the collection is a set. So if the collection is a set, you need to change it back to values() - which is the only thing that can be indexed in a set.

Other than that, this PR looks good!

I didn't like this "target rewrite" thing when I reviewed the collection indexing code, and now I remember why :-(

piodul commented:
Setting the `backport/6.0` label because it's broken there as well.

piodul commented on test/cql-pytest/test_describe.py:
```diff
@@ -447,6 +447,24 @@ def test_desc_index(cql, test_keyspace):
+            ("idx_c_keys",      f"CREATE INDEX idx_c_keys ON {tbl}(keys(c))"),
+            ("idx_c_values",    f"CREATE INDEX idx_c_values ON {tbl}(values(c))"),
+            ("idx_d",           f"CREATE INDEX idx_d ON {tbl}(d)"),    # TODO: this index will fail the assert becuase its target is "values(d)"
+            ("idx_e",           f"CREATE INDEX idx_e ON {tbl}(e)"),    # TODO: this index will fail the assert becuase its target is "keys(e)"
```
Nit: ditto

piodul replied:
By the way @tchaikov, is there a reason why we don't run codespell on test files? I can see that the `test` directory is on the skip list in `.github/workflows/codespell.yaml`.

tchaikov replied:
because i didn't have enough bandwidth to address all the issues found in `test` directory. there are still 52 of them:
```console
$ codespell test -L ans,datas,fo,ser,ue,crate,nd,reenable,strat,stap,te,raison | wc -l
52
```

once we fix all of them, we should be able to remove `test` from the skip list.

tchaikov replied:
i took this opportunity to fix some of them, see #19412

Jadw1 commented:
Changelog:
- fixed problem with index on set
- extended the test to check all possibilities

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9878/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9878/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: cql-pytest/test_describe
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9878/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2206/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10229/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9878/)
#### Failed Tests (13/41395):
- [test_drop_index_while_building](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9878/testReport/junit/secondary_indexes_test/TestSecondaryIndexes/Tests___dtest_with_topology_changes___test_drop_index_while_building) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_drop_index_while_building)
- [test_drop_index_while_building](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9878/testReport/junit/secondary_indexes_test/TestSecondaryIndexes/Tests___dtest_with_topology_changes___test_drop_index_while_building_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_drop_index_while_building)
- [test_drop_index_while_building](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9878/testReport/junit/secondary_indexes_test/TestSecondaryIndexes/Tests___dtest___test_drop_index_while_building) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_drop_index_while_building)
- [test_drop_index_while_building](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9878/testReport/junit/secondary_indexes_test/TestSecondaryIndexes/Tests___dtest___test_drop_index_while_building_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_drop_index_while_building)
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9878/testReport/junit/boost.secondary_index_test.test_secondary_index_name.1.xunit/xml/Tests___Unit_Tests____empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9878/testReport/junit/boost.secondary_index_test.test_secondary_index_name.1.xunit/xml/Tests___Unit_Tests____empty__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9878/testReport/junit/boost.secondary_index_test.test_secondary_index_name.1.xunit/xml/Tests___Unit_Tests____empty__3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9878/testReport/junit/boost.secondary_index_test.test_secondary_index_name.2.xunit/xml/Tests___Unit_Tests____empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9878/testReport/junit/boost.secondary_index_test.test_secondary_index_name.2.xunit/xml/Tests___Unit_Tests____empty__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9878/testReport/junit/boost.secondary_index_test.test_secondary_index_name.2.xunit/xml/Tests___Unit_Tests____empty__3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
#### Build Details:
- Duration: 2 hr 36 min
- Builder: spider8.cloudius-systems.com

tchaikov commented on index/secondary_index.cc:
```diff
@@ -106,6 +106,16 @@ sstring target_parser::get_target_column_name_from_string(const sstring& targets
+
+    auto opening_pos = target.find("(");
+    auto closing_pos = target.find(")");
+    return target.substr(opening_pos + 1, closing_pos - opening_pos - 1);
```
this looks for the `(` and `)` characters twice. how about
```c++
auto opening_pos = target.find("(");
if (opening_pos == target.npos) {
    return target;
}
auto closing_pos = target.find(")", ++opening_pos);
if (closing_pos == target.npos) {
    return target;
}
return target.substr(opening_pos, closing_pos - opening_pos);
```

Jadw1 commented:
Changelog:
- The tests were failing because I didn't take quoting into account. Fixed the problem
- added a test for quoting

Jadw1 commented:
Ping @nyh @piodul

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10017/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10017/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: cql-pytest/test_describe
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10017/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2307/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10017/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10325/)
#### Build Details:
- Duration: 3 hr 12 min
- Builder: spider7.cloudius-systems.com

piodul commented on index/secondary_index.cc:
```diff
@@ -106,6 +106,26 @@ sstring target_parser::get_target_column_name_from_string(const sstring& targets
     return targets;
 }
 
+sstring target_parser::get_base_column_name_from_target(const sstring& target) {
```
Nit: always consider using a `std::string_view` instead of `sstring`/`std::string` if you pass a string by reference. It's just more flexible (accepts both `sstring` and `std::string`, also other `std::string_view`) and more performant (no need to materialize a string object, e.g. when passing a string literal).

piodul commented on index/secondary_index.cc:
```diff
@@ -106,6 +106,26 @@ sstring target_parser::get_target_column_name_from_string(const sstring& targets
+    }
+
+    // unquote column name if it's quoted
+    if (base_name[0] == '"' && base_name[base_name.length() - 1] == '"') {
```
What if `base_name` is empty? Not sure if it can happen at all, but I would feel more confident about this code if this case were properly handled. This also doesn't seem to work if the string is just a single `"` (you will try to call substr with negative length).

piodul replied:
If the string might be quoted then the next natural question is, "can this string contain some escaped characters"?

Which code/function is responsible for adding the quotes? Is there some unquote logic/function that exists and could be used here?

Jadw1 replied:
> Which code/function is responsible for adding the quotes? Is there some unquote logic/function that exists and could be used here?

`cql3::util::maybe_quote()` and it contains logic to replace `"` in the name with `""`. But AFAIK it cannot contain other escaped characters.
@nyh you are the expert of cql

nyh replied:
Yes, we have the function index_target::unescape_target_column() and you should use it.
It was introduced in https://github.com/scylladb/scylladb/pull/10841/commits/2c244c6e09c4693a47eadded930651599f8145d1 which also explains why this quoting was needed.

piodul commented on index/secondary_index.cc:
```diff
@@ -106,6 +106,26 @@ sstring target_parser::get_target_column_name_from_string(const sstring& targets
+    if (opening_pos != target.npos) {
+        auto closing_pos = target.find(")", ++opening_pos);
+        if (opening_pos != target.npos) {
+            base_name = target.substr(opening_pos, closing_pos - opening_pos);
```
This operation would also be more efficient if `base_name`/`target` were `std::string_views`.

piodul commented on test/cql-pytest/test_describe.py:
```diff
@@ -447,6 +447,49 @@ def test_desc_index(cql, test_keyspace):
+            ("idx_d_keys",      f"CREATE INDEX idx_d_keys ON {tbl}(keys(d))"),
+            ("idx_d_values",    f"CREATE INDEX idx_d_values ON {tbl}(values(d))"),
+            ("idx_e",           f"CREATE INDEX idx_e ON {tbl}(values(e))"),
+            ("idx_f",           f"CREATE INDEX idx_f ON {tbl}(values(f))"),
```
Can we include some cases that would exercise the "unquoting" logic?

Jadw1 replied:
Cases with unquoting are in another test in this commit (test_desc_index_column_quoting)

piodul replied:
Oh, right. Sorry, I missed them. Nonetheless, more test cases should be added to cover strings with escaped characters.

nyh replied:
See also the tests in test/cql-pytest/test_secondary_index.py, test_index_quoted_names, which checks all the interesting edge cases of the quoting (not describe table of course, just the create index).

The most "difficult" edge case is a column name like `keys(m)` which looks like it's a collection, but isn't. This case is handled correctly, and tested by that test and we should also test it in the context of DESCRIBE.

Jadw1 replied:
copied the cases from the test @nyh  pointed

nyh commented:
I plan to review this PR now, but I see there were some questions on how the index "target" was saved in the system table. Unfortunately, because of historic reasons, it's a mess. See commit https://github.com/scylladb/scylladb/pull/10841/commits/2c244c6e09c4693a47eadded930651599f8145d1 on the three formats it can have - column name, keys(colname) and JSON string. That column also introduces the necessary escaping of the column name in the first case, and functions to do this escaping and unescaping (index_target::unescape_target_column)

nyh commented on index/target_parser.hh:
```diff
@@ -29,6 +29,8 @@ struct target_parser {
 
     static sstring get_target_column_name_from_string(const sstring& targets);
 
+    static sstring get_base_column_name_from_target(const sstring& target);
```
I don't know what a "base column name" means and how it differs from "target column name". As https://github.com/scylladb/scylladb/pull/10841/commits/2c244c6e09c4693a47eadded930651599f8145d1 explains, we have three options now - a single column name (maybe quoted), a string(columname) (e.g., keys(colname), and a JSON (for "local index"). Which of these cases does this new function support and why?

nyh replied:
Maybe you are aiming for something like the existing function index_target::column_name_from_target_string()?

Jadw1 replied:
Yes, exactly this. Thanks!

nyh commented on index/secondary_index.cc:
```diff
@@ -106,6 +106,26 @@ sstring target_parser::get_target_column_name_from_string(const sstring& targets
+    sstring base_name = target;
+    
+    // extract column name if the target is in format "function(column_name)"
+    auto opening_pos = target.find("(");
```
You need to *first* check if the target begins with a quote, because if it does it means this is a column name, even if it contains the "(" character (which is allowed in a column name).
I think index_target::column_name_from_target_string already does this correctly.
Please see commit https://github.com/scylladb/scylladb/pull/10841/commits/2c244c6e09c4693a47eadded930651599f8145d1 for an explanation of the different target encodings.

nyh commented on schema/schema.cc:
```diff
@@ -821,6 +822,48 @@ sstring schema::element_type(replica::database& db) const {
     return "table";
 }
 
+static void describe_index_columns(std::ostream& os, bool is_local, const schema& index_schema, schema_ptr base_schema) {
```
There's something I don't understand here, but I have to be honest, I don't remember the details of what we can or can't be done.

The target **strings** are meant to be used in our on-disk format. But in memory, we have schema and index_manager. Those should give you objects with _index.target_type() and target_column(), so you shouldn't need to process target strings at all... Are you sure you needed to do this?

Jadw1 replied:
Yes, I saw those functions but they are methods of `cql3::statements::index_target` class. And this class is only used when creating the index and AFAIK it's not used or stored anywhere else.

And to recreate `index_target` objects myself, I would need to parse target string to extract target type and column definition, so there is no point to create the object only to get the arguments I've just used to create it.

nyh commented on test/cql-pytest/test_describe.py:
```diff
@@ -447,6 +447,49 @@ def test_desc_index(cql, test_keyspace):
         
         assert f"CREATE INDEX named_index ON {tbl}{maybe_space}(c)" in c_desc
 
+def test_desc_index_on_collections(cql, test_keyspace):
```
Does this test pass on Cassandra as well? If not, please think if it shouldn't, and consider adding the appropriate tag (xfail/scylla_only/cassandra_bug) depending on why it works differently in Cassandra.

Jadw1 replied:
Yes, it passes but it required a small change in assertions (see comment in the code).

In short: Scylla outputs `CREATE INDEX ON table_name(column_name)`  but Cassandra does `CREATE INDEX ON table_name (column_name)` (extra whitespace if you didn't noticed :D )

nyh commented on test/cql-pytest/test_describe.py:
```diff
@@ -447,6 +447,49 @@ def test_desc_index(cql, test_keyspace):
+    with new_test_table(cql, test_keyspace, "a int primary key, \"select\" int, \"some???name\" map<int, int>") as tbl:
+        indexes = [
+            ("idx_1", f"CREATE INDEX idx_1 ON {tbl}(\"select\")"),
+            ("idx_2", f"CREATE INDEX idx_2 ON {tbl}(keys(\"some???name\"))"),
```
As I mentioned above, the craziest edge case is a column whose name is "keys(m)" and you can index this column just fine, and it needs to work :-) Please add that edge case, I had to work hard for it to actually work in https://github.com/scylladb/scylladb/pull/10841/commits/2c244c6e09c4693a47eadded930651599f8145d1 :-)

nyh commented on test/cql-pytest/test_describe.py:
```diff
@@ -447,6 +447,49 @@ def test_desc_index(cql, test_keyspace):
+        for name, create_stmt in indexes:
+            cql.execute(create_stmt)
+            desc = cql.execute(f"DESC INDEX {test_keyspace}.{name}").one().create_statement
+            assert create_stmt in desc
```
I think (but correct me if I'm wrong) you don't have a test for local secondary index of collection, so please add it. Of course this will be `scylla_only` because local secondary index is a Scylla extension that doesn't exist in Cassandra.

Jadw1 replied:
Added test for secondary indexes on regular column and on frozen collection column. Indexes on other column types like: set<>, list<> are not supported yet.

Also, I didn't created a test for quoting in secondary indexes because the code to generate description in Scylla is exactly the same.

Jadw1 commented:
Changelog:
- dropped the first commit and used `index_target::column_name_from_target_string()` function instead, since it's doing exactly what I need
- fixed the code to also properly describe secondary indexes
- extended the tests to check quoting in column names and to check describe of secondary indexes

nyh commented:
>     * fixed the code to also properly describe secondary indexes

Your original patch was about describing secondary indexes, so what changed here? Maybe you meant to write "local secondary indexes" and forgot the word "local"?

Jadw1 commented:
> > ```
> > * fixed the code to also properly describe secondary indexes
> > ```
> 
> Your original patch was about describing secondary indexes, so what changed here? Maybe you meant to write "local secondary indexes" and forgot the word "local"?

Right. Updated the changelog and I'll update test name in a second

nyh commented on test/cql-pytest/test_describe.py:
```diff
@@ -447,6 +447,75 @@ def test_desc_index(cql, test_keyspace):
 
+def test_desc_index_on_collections(cql, test_keyspace):
+    # In this test, all assertions are in form of
+    # `assert create_stmt in desc or create_stmt.replace("(", " (", 1)`
```
Maybe I'm misreading something, but isn't the "or" part an error? Didn't you for get the "in desc" again? :-(

Jadw1 replied:
Right, I missed that.

nyh commented on test/cql-pytest/test_describe.py:
```diff
@@ -447,6 +447,75 @@ def test_desc_index(cql, test_keyspace):
+    #   Scylla:     `CREATE INDEX ON table_name(column_name)`
+    #   Cassandra:  `CREATE INDEX ON table_name (column_name)`
+
+    quoted_names = ['"select"', '"hEllo"', '"x y"', '"hi""hello""yo"', '"""hi"""', '"keys(m)"', '"values(m)"', '"entries(m)"']
```
I see you did check the case of a column named `keys(m)` but this column is not a list (it's an int) so it doesn't catch the bug I suspected in the previous patch. So I guess if this is a real bug you'll need a separate test for it.

Jadw1 replied:
changed the test so columns with names keys/values/entries have appropriate types

nyh commented on test/cql-pytest/test_describe.py:
```diff
@@ -447,6 +447,75 @@ def test_desc_index(cql, test_keyspace):
+        for name, create_stmt in indexes:
+            cql.execute(create_stmt)
+            desc = cql.execute(f"DESC INDEX {test_keyspace}.{name}").one().create_statement
+            assert create_stmt in desc
```
Why is this "in" and not equal?

Jadw1 replied:
I was doing assertions this way in the whole test file. I could also do `assert create_stmt == desc.trim()` in most places but I just left it this way.

nyh reviewed (changes requested):
Thanks, looks mostly good but I suspect a few small remaining problems so please have a look at my comments. Thanks.

Jadw1 commented on schema/schema.cc:
```diff
@@ -821,6 +822,48 @@ sstring schema::element_type(replica::database& db) const {
     return "table";
 }
 
+static void describe_index_columns(std::ostream& os, bool is_local, const schema& index_schema, schema_ptr base_schema) {
```
Just for readability reasons. `describe()` method is used to describe tables, indexes and MVs, so it seemed reasonable to extract code which is responsible for describing index columns.

Jadw1 commented on schema/schema.cc:
```diff
@@ -821,6 +822,48 @@ sstring schema::element_type(replica::database& db) const {
+        index_metadata.options().at(cql3::statements::index_target::target_option_name)
+    );
+    auto base_column_name = cql3_parser::index_target::column_name_from_target_string(target_str);
+    auto base_column = base_schema->get_column_definition(to_bytes(base_column_name));
```
added a check to throw on_internal_error

Jadw1 commented on schema/schema.cc:
```diff
@@ -821,6 +822,48 @@ sstring schema::element_type(replica::database& db) const {
+    if (bk_type->is_collection() && !bk_type->is_multi_cell()) {
+        // Indexes on frozen collection require full() function but the function is dropped while saving the target.
+        os << "full(" << cql3::util::maybe_quote(base_column_name) << ")";
+    } else if (bk_type->is_set() && target_str.contains("keys")) {
```
Changed to check prefix of the name and added the test case

Jadw1 commented on schema/schema.cc:
```diff
@@ -821,6 +822,48 @@ sstring schema::element_type(replica::database& db) const {
+    // Global and local indexes may only contain one column.
+    // Local indexes support multi column indexes via custom indexes, but at least
+    // for now, Scylla supports no custom indexes.
+    auto index_metadata = base_schema->all_indices().at(secondary_index::index_name_from_table_name(index_schema.cf_name()));
```
It shouldn't happen, the function is executed only when a schema in describe() method is an index. But I've added a check and extended the test to try to describe non existing index and `DESC INDEX table_name`

Jadw1 commented:
Changelog:
- addressed Nadav comments: added sanity checks, extended comments, looking for `keys(` prefix instead of checking if string contains `keys` substring (in describe_index_columns method)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10100/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10100/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: cql-pytest/test_describe
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10100/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10395/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2370/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10100/)
#### Build Details:
- Duration: 3 hr 9 min
- Builder: spider5.cloudius-systems.com

piodul commented on test/cql-pytest/test_describe.py:
```diff
@@ -447,6 +447,83 @@ def test_desc_index(cql, test_keyspace):
+        cql.execute(f"CREATE INDEX idx_g ON {tbl}(g)")
+        idx_g_desc = cql.execute(f"DESC INDEX {test_keyspace}.idx_g").one().create_statement
+        idx_g_create_stmt = f"CREATE INDEX idx_g ON {tbl}(values(g))"
+        assert idx_g_create_stmt in idx_g_desc or idx_g_create_stmt.replace("(", " (", 1)
```
I'm afraid `assert x in y or z` does not do what you think it does - to be fair, I'm not 100% sure what was your intention, but the way it works looks wrong. It's parsed as `assert (x in y) or z`. For example, this code does not trigger an assert:

```python
assert 0 in [1] or [2]
```

Python has a concept of "truthiness" and some values can be interpreted as booleans even though they aren't. In your case for example, non-empty strings (like `idx_g_create_stmt.replace("(", " (", 1)`) are always truthy. I'm afraid this assertion will always pass because the string at the end will be non-empty.

Was this supposed to be

```python
assert idx_g_create_stmt in idx_g_desc \
        or idx_g_create_stmt.replace("(", " (", 1) in idx_g_desc
```

?

Lots of assertions in the tests that you introduce have this problem.

piodul replied:
I recommend introducing a helper "normalize" function which gets rid of the space before the parenthesis (or adds it, but only if there was no space there in the first place) and then use the `in` operator on those.

Jadw1 replied:
Yes, it wasn't intentional. I've simply made a mistake.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10307/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10307/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: cql-pytest/test_describe
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10307/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2529/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10554/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10307/)
#### Build Details:
- Duration: 2 hr 53 min
- Builder: spider3.cloudius-systems.com

Jadw1 commented:
Changelog:
- fixed the mistake in the test

piodul commented:
Queued
